Configure packages on Windows no longer work with autoconf >= 2.70

**Describe the bug**
Ever since autoconf 2.70 all configure based packages are broken on Windows with

```
configure: error: cannot find required auxiliary files: config.guess config.sub
```

(This is why I hate configure based packages). This autoconf introduced a change
in how it finds auxilary tools/scripts [4].

This breaks packages like `network` and `ghc` itself [1][2].

There's un upstream bug-report [3] but this is unlikely to be fixed.

A number of things go wrong here:

Cabal calls configure with Windows paths as arguments, e.g.:

```
"C:\tools\msys64\usr\bin\sh.exe" "C:\/Users/xbox/source/repos/network/configure"
```

But because the default PATH separator is `:`, so autoconf is interpeting this as:

```
C/
\/Users/xbox/source/repos/network/configure
```

Both of which are incorrect. `cabal` should be passing a Unix path here as `configure`
is a unix tool.  Passing a unix path won't entirely fix the issue however
because during the call a couple of ENV variables are passed to `configure` which contain
Windows paths such as:

```
"ACLOCAL_PATH","C:\\tools\\msys64\\mingw64\\share\\aclocal;C:\\tools\\msys64\\usr\\share\\aclocal"
```

So the easiest fix for both issues is that cabal sets the environment variable:

```
env PATH_SEPARATOR=";" 
```

before the call to `sh`.

**To Reproduce**
Steps to reproduce the behavior:

On a checked out version of `network` or any other configure based package

```
$ cabal build --enable-tests --disable-benchmarks all
```

**Expected behavior**
For configure packages to work. We'll likely need a release of `cabal-install` for this.

**System information**
 - Windows 10
 - `cabal`, version 3.2.0.0

**Additional context**
[1] https://gitlab.haskell.org/ghc/ghc/-/issues/19189
[2] https://github.com/haskell/network/issues/502
[3] https://savannah.gnu.org/support/index.php?110448
[4] http://git.savannah.gnu.org/gitweb/?p=autoconf.git;a=blob_plain;f=NEWS;hb=97fbc5c184acc6fa591ad094eae86917f03459fa

--- a/Cabal/src/Distribution/Simple.hs
+++ b/Cabal/src/Distribution/Simple.hs
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE LambdaCase #-}
@@ -715,7 +716,11 @@ runConfigureScript verbosity backwardsCompatHack flags lbi = do
       pathEnv = maybe (intercalate spSep extraPath)
                 ((intercalate spSep extraPath ++ spSep)++) $ lookup "PATH" env
       overEnv = ("CFLAGS", Just cflagsEnv) :
+-- TODO: Move to either Cabal/src/Distribution/Compat/Environment.hs
+-- or Cabal/src/Distribution/Compat/FilePath.hs:
+#ifdef mingw32_HOST_OS
                 ("PATH_SEPARATOR", Just ";") :
+#endif
                 [("PATH", Just pathEnv) | not (null extraPath)]
       hp = hostPlatform lbi
       maybeHostFlag = if hp == buildPlatform then [] else ["--host=" ++ show (pretty hp)]

--- a/Cabal/src/Distribution/Simple.hs
+++ b/Cabal/src/Distribution/Simple.hs
@@ -715,6 +715,7 @@ runConfigureScript verbosity backwardsCompatHack flags lbi = do
       pathEnv = maybe (intercalate spSep extraPath)
                 ((intercalate spSep extraPath ++ spSep)++) $ lookup "PATH" env
       overEnv = ("CFLAGS", Just cflagsEnv) :
+                ("PATH_SEPARATOR", Just ";") :
                 [("PATH", Just pathEnv) | not (null extraPath)]
       hp = hostPlatform lbi
       maybeHostFlag = if hp == buildPlatform then [] else ["--host=" ++ show (pretty hp)]

--- a/changelog.d/pr-7510
+++ b/changelog.d/pr-7510
@@ -1,0 +1,4 @@
+synopsis: Set PATH_SEPARATOR=";" when calling ./configure on Windows; this fix is necessary for autoconf >= 2.70
+packages: Cabal
+prs: #7510
+issues: #7494

